src/sinol_make/commands/gen/__init__.py

```diff
@@ -0,0 +1,147 @@
+                                 f'Used when generating output files.', default=mp.cpu_count())
+        parsers.add_compilation_arguments(parser)
+
+    def compile_ingen(self):
+        self.ingen_exe, compile_log_path = gen_util.compile_ingen(self.ingen, self.args,
+                                                                  self.args.weak_compilation_flags)
+        if self.ingen_exe is None:
+            print(util.error('Failed ingen compilation.'))
+            compile.print_compile_log(compile_log_path)
+            exit(1)
+        else:
+            print(util.info('Successfully compiled ingen.'))
+
+    def compile_correct_solution(self):
+        self.correct_solution_exe, compile_log_path = gen_util.compile_correct_solution(self.correct_solution, self.args,
+                                                                                        self.args.weak_compilation_flags)
+        if self.correct_solution_exe is None:
+            print(util.error('Failed ingen compilation.'))
+            compile.print_compile_log(compile_log_path)
+            exit(1)
+        else:
+            print(util.info('Successfully compiled ingen.'))
+
+    def generate_outputs(self, outputs_to_generate):
+        print(f'Generating output files for {len(outputs_to_generate)} tests on {self.args.cpus} cpus.')
+        arguments = []
+        for output in outputs_to_generate:
+            output_basename = os.path.basename(output)
+            input = os.path.join(os.getcwd(), 'in', os.path.splitext(output_basename)[0] + '.in')
+            arguments.append(OutputGenerationArguments(self.correct_solution_exe, input, output))
+
+        failed = False
+        with mp.Pool(self.args.cpus) as pool:
+            for i, result in enumerate(pool.imap_unordered(gen_util.generate_output, arguments)):
+                if result:
+                    print(util.info(f'Successfully generated output file {os.path.basename(arguments[i].output_test)}'))
+                else:
+                    failed = True
+                    print(util.error(f'Failed to generate output file {os.path.basename(arguments[i].output_test)}'))
+
+        if failed:
+            util.exit_with_error('Failed to generate some output files.')
+        else:
+            print(util.info('Successfully generated all output files.'))
+
+    def calculate_md5_sums(self):
+        """
+        Calculates md5 sums for each test.
+        :return: Tuple (dictionary of md5 sums, list of outputs tests that need to be generated)
+        """
+        old_md5_sums = None
+        if os.path.exists(os.path.join(os.getcwd(), 'in', '.md5sums')):
+            try:
+                with open(os.path.join(os.getcwd(), 'in', '.md5sums'), 'r') as f:
+                    old_md5_sums = yaml.load(f, Loader=yaml.FullLoader)
+            except yaml.YAMLError:
+                pass
+            if not isinstance(old_md5_sums, dict):
+                old_md5_sums = None
```

Comment:
This code's flow is a bit chaotic. Can it be made better?

Reply:
I couldn't come up with anyting better

Reply:
Maybe except for adding `md5_file = os.path.join(os.getcwd(), 'in', '.md5sums')`

Reply:
Hmm, maybe checking whether the file exists and reading it can be combined into one?

Reply:
```
old_md5_sums = None
try:
    with open(os.path.join(os.getcwd(), 'in', '.md5sums'), 'r') as f:
        if isinstance(old_md5_sums, dict):
            old_md5_sums = yaml.load(f, Loader=yaml.FullLoader)
except smth:
    pass
```
(hand-written, it might not work)

Reply:
I changed it, now it looks similar to what you wrote